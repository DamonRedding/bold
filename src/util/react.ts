import { MutableRefObject, Ref } from 'react'

export function composeRefs<T>(...refs: Array<Ref<T>>) {
    return (value: T) => refs.forEach(ref => setRef(ref, value))
}

export function setRef<T>(ref: Ref<T>, value: any) {
    if (typeof ref === 'function') {
        ref(value)
    } else if (ref) {
        (ref as MutableRefObject<T>).current = value
    }
}

/**
 * Returns the next sibling of `element` that matches the `predicate`.
 * Returns `null` if not found.
 */
export function getNextSibling(element: Element, predicate: (sibling: Element) => boolean) {
    let current: Element = element

    while (current.nextElementSibling) {
        current = current.nextElementSibling

        if (predicate(current)) {
            return current
        }
    }

    return null
}

/**
 * Returns the previous sibling of `element` that matches the `predicate`.
 * Returns `null` if not found.
 */
export function getPreviousSibling(element: Element, predicate: (sibling: Element) => boolean) {
    let current: Element = element

    while (current.previousElementSibling) {
        current = current.previousElementSibling

        if (predicate(current)) {
            return current
        }
    }

    return null
}
