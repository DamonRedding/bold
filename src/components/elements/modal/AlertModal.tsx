import * as React from 'react'
import * as classnames from 'classnames'
import { Modal } from '../Modal'
import { Icon } from '../Icon'
import { Button } from '../Button/Button'
import { Linha } from '../../layout/Linha'
import { AlertModalButtonBar } from './AlertModalButtonBar'
import { AlertModalContent } from './AlertModalContent'

export interface AlertModalProps {
    active?: boolean
    error?: boolean
    icon?: string
    info?: boolean
    success?: boolean
}

export class AlertModal extends React.Component<AlertModalProps, any> {

    getIcon = () => {
        if (this.props.icon) {
            return this.props.icon
        }
        if (this.props.success) {
            return 'modal-sucesso'
        } else if (this.props.error) {
            return 'modal-erro'
        } else if (this.props.info) {
            return 'modal-atencao'
        }
        return ''
    }

    render() {
        const tipo = classnames('alertmodal-card modal-content', {
            ' is-success': this.props.success,
            ' is-error': this.props.error,
            ' is-info': this.props.info
        })

        return (
            <Modal active={this.props.active}>
                <div className={tipo}>
                    <Icon icon={this.getIcon()} className='alertmodal-icon' />
                    {this.props.children}
                </div>
            </Modal>
        )
    }

}

export interface AlertModalErrorProps {
    active?: boolean
    error?: any
    icon?: string
    onClose: () => void
    title?: string
}

export const AlertModalError: React.SFC<AlertModalErrorProps> = (props) => {

    return (
        <AlertModal active={props.active} icon={props.icon} error>
            <AlertModalContent title={props.title}>
                {props.children}
            </AlertModalContent>
            <AlertModalButtonBar>
                <Button type='primary' size='medium' onClick={props.onClose}>OK</Button>
            </AlertModalButtonBar>
        </AlertModal>
    )

}

export interface AlertModalLeaveProps {
    active?: boolean
    callback: (shouldNavigate: boolean) => void
}

export class AlertModalLeave extends React.Component<AlertModalLeaveProps, any> {

    constructor(props, context?) {
        super(props, context)

        this.state = {
            active: true
        }
    }

    componentWillReceiveProps() {
        this.state = {
            active: true
        }
    }

    onCancel = () => {
        this.props.callback(false)
        this.close()
    }

    onLeave = () => {
        this.props.callback(true)
        this.close()
    }

    close() {
        this.setState({ active: false })
    }

    render() {
        return (
            <AlertModal active={this.state.active} info>
                <AlertModalContent title='Deseja sair?'>
                    Você preencheu alguns dados, se você sair agora esses dados serão perdidos. Tem certeza que deseja sair?
                </AlertModalContent>
                <AlertModalButtonBar >
                    <Button
                        size='medium'
                        onClick={this.onCancel}>Cancelar</Button>
                    <Button
                        type='primary'
                        size='medium'
                        onClick={this.onLeave}>Sair</Button>
                </AlertModalButtonBar>
            </AlertModal>
        )
    }

}

export interface AlertModalSuccessProps {
    active?: boolean
    icon?: string
    onClose: () => void
    result?: any
    title?: string
}

export const AlertModalSuccess: React.SFC<AlertModalSuccessProps> = (props) => {
    return (
        <AlertModal active={props.active} icon={props.icon} success>
            <AlertModalContent title={props.title}>
                {props.children}
            </AlertModalContent>
            <AlertModalButtonBar >
                <Linha />
                <Button
                    type='success'
                    size='medium'
                    onClick={props.onClose}>OK</Button>
            </AlertModalButtonBar>
        </AlertModal>
    )
}

export interface AlertModalDeleteProps {
    active?: boolean
    onClose: () => void
    onOk: () => void
}

export const AlertModalDelete: React.SFC<AlertModalDeleteProps> = (props) => {
    return (
        <AlertModalConfirmacao {...props} title='Confirmar exclusão!'>
            Tem certeza que deseja excluir esse item?
        </AlertModalConfirmacao>
    )
}

export interface AlertModalConfirmacaoProps {
    active?: boolean
    icon?: string
    onClose: () => void
    onOk: () => void
    title?: string
}

export const AlertModalConfirmacao: React.SFC<AlertModalConfirmacaoProps> = (props) => {
    return (
        <AlertModal active={props.active} icon={props.icon} info>
            <AlertModalContent title={props.title}>
                {props.children}
            </AlertModalContent>
            <AlertModalButtonBar>
                <Button size='medium' onClick={props.onClose}>Cancelar</Button>
                <Button type='primary' size='medium' onClick={props.onOk}>Confirmar</Button>
            </AlertModalButtonBar>
        </AlertModal>
    )
}
