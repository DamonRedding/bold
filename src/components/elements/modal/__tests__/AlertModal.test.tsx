import * as React from 'react'
import { shallowRenderAndMatch } from '../../../../__tests__/test.defaults'
import { AlertModalError, AlertModalLeave, AlertModalSuccess, AlertModalDelete, AlertModalConfirmacao } from '../AlertModal'

describe('AlertModal', () => {
    it('deve renderizar AlertModalError visível', function () {
        shallowRenderAndMatch(<AlertModalError active={true} title='AlertModalError' onClose={() => undefined}>Descrição</AlertModalError>)
    })
    it('deve renderizar AlertModalError', function () {
        shallowRenderAndMatch(<AlertModalError title='AlertModalError' onClose={() => undefined}>Descrição</AlertModalError>)
    })
    it('deve renderizar AlertModalLeave visível', function () {
        shallowRenderAndMatch(<AlertModalLeave callback={() => undefined} />)
    })
    it('deve renderizar AlertModalSuccess visível', function () {
        shallowRenderAndMatch(<AlertModalSuccess active={true} title='Sucesso' onClose={() => undefined}>Descrição</AlertModalSuccess>)
    })
    it('deve renderizar AlertModalSuccess', function () {
        shallowRenderAndMatch(<AlertModalSuccess title='Sucesso' onClose={() => undefined}>Descrição</AlertModalSuccess>)
    })
    it('deve renderizar AlertModalDelete visível', function () {
        shallowRenderAndMatch(<AlertModalDelete active={true} onOk={() => undefined} onClose={() => undefined} />)
    })
    it('deve renderizar AlertModalDelete', function () {
        shallowRenderAndMatch(<AlertModalDelete onOk={() => undefined} onClose={() => undefined} />)
    })
    it('deve renderizar AlertModalConfirmacao visível', function () {
        shallowRenderAndMatch(<AlertModalConfirmacao active={true} title='Sucesso' onOk={() => undefined} onClose={() => undefined}>Descrição</AlertModalConfirmacao>)
    })
    it('deve renderizar AlertModalConfirmacao', function () {
        shallowRenderAndMatch(<AlertModalConfirmacao title='Sucesso' onOk={() => undefined} onClose={() => undefined}>Descrição</AlertModalConfirmacao>)
    })
})
