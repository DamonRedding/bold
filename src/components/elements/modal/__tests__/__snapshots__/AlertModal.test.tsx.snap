// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AlertModal deve renderizar AlertModalConfirmacao 1`] = `
<AlertModal
  info={true}
>
  <Component
    title="Sucesso"
  >
    Descrição
  </Component>
  <Button
    label="Cancelar"
    onClick={[Function]}
  />
  <Button
    label="Confirmar"
    onClick={[Function]}
    type="primary"
  />
</AlertModal>
`;

exports[`AlertModal deve renderizar AlertModalConfirmacao visível 1`] = `
<AlertModal
  active={true}
  info={true}
>
  <Component
    title="Sucesso"
  >
    Descrição
  </Component>
  <Button
    label="Cancelar"
    onClick={[Function]}
  />
  <Button
    label="Confirmar"
    onClick={[Function]}
    type="primary"
  />
</AlertModal>
`;

exports[`AlertModal deve renderizar AlertModalDelete 1`] = `
<Component
  onClose={[Function]}
  onOk={[Function]}
  title="Confirmar exclusão!"
>
  Tem certeza que deseja excluir esse item?
</Component>
`;

exports[`AlertModal deve renderizar AlertModalDelete visível 1`] = `
<Component
  active={true}
  onClose={[Function]}
  onOk={[Function]}
  title="Confirmar exclusão!"
>
  Tem certeza que deseja excluir esse item?
</Component>
`;

exports[`AlertModal deve renderizar AlertModalError 1`] = `
<AlertModal
  error={true}
>
  <Component
    title="AlertModalError"
  >
    Descrição
  </Component>
  <Button
    label="OK"
    onClick={[Function]}
    type="primary"
  />
</AlertModal>
`;

exports[`AlertModal deve renderizar AlertModalError visível 1`] = `
<AlertModal
  active={true}
  error={true}
>
  <Component
    title="AlertModalError"
  >
    Descrição
  </Component>
  <Button
    label="OK"
    onClick={[Function]}
    type="primary"
  />
</AlertModal>
`;

exports[`AlertModal deve renderizar AlertModalLeave visível 1`] = `
<AlertModal
  active={true}
  info={true}
>
  <Component
    title="Deseja sair?"
  >
    Você preencheu alguns dados, se você sair agora esses dados serão perdidos. Tem certeza que deseja sair?
  </Component>
  <Button
    label="Cancelar"
    onClick={[Function]}
  />
  <Button
    label="Sair"
    onClick={[Function]}
    type="primary"
  />
</AlertModal>
`;

exports[`AlertModal deve renderizar AlertModalSuccess 1`] = `
<AlertModal
  success={true}
>
  <Component
    title="Sucesso"
  >
    Descrição
  </Component>
  <Button
    label="OK"
    onClick={[Function]}
  />
</AlertModal>
`;

exports[`AlertModal deve renderizar AlertModalSuccess visível 1`] = `
<AlertModal
  active={true}
  success={true}
>
  <Component
    title="Sucesso"
  >
    Descrição
  </Component>
  <Button
    label="OK"
    onClick={[Function]}
  />
</AlertModal>
`;
